lib/hashtable.cc

```diff
@@ -242,327 +241,115 @@ bool Hashtable::median_at_least(const std::string &s,
 
     while(!kmers.done()) {
-        kmer = kmers.next();
-        bool is_new_kmer;
-
-        // Set the bits for the kmer in the various hashtables,
-        // and report on whether or not they had already been set.
-        // This is probably better than first testing and then setting the bits,
-        // as a failed test essentially results in doing the same amount of work
-        // twice.
-        if ((is_new_kmer = store->test_and_set_bits( kmer ))) {
-            ++n_consumed;
-        }
-
-#if (1)
-        if (is_new_kmer) {
-            ++since;
-        } else {
-            ACQUIRE_ALL_TAGS_SPIN_LOCK
-            kmer_tagged = set_contains(all_tags, kmer);
-            RELEASE_ALL_TAGS_SPIN_LOCK
-            if (kmer_tagged) {
-                since = 1;
-                if (found_tags) {
-                    found_tags->insert(kmer);
-                }
-            } else {
-                ++since;
-            }
-        }
-#else
-        if (!is_new_kmer && set_contains(all_tags, kmer)) {
-            since = 1;
-            if (found_tags) {
-                found_tags->insert(kmer);
-            }
-        } else {
-            since++;
-        }
-#endif
-
-        if (since >= _tag_density) {
-            ACQUIRE_ALL_TAGS_SPIN_LOCK
-            all_tags.insert(kmer);
-            RELEASE_ALL_TAGS_SPIN_LOCK
-            if (found_tags) {
-                found_tags->insert(kmer);
-            }
-            since = 1;
-        }
+        HashIntoType kmer = kmers.next();
 
-    } // iteration over kmers
+        BoundedCounterType count = this->get_count(kmer);
 
-    if (since >= _tag_density/2 - 1) {
-        ACQUIRE_ALL_TAGS_SPIN_LOCK
-        all_tags.insert(kmer);	// insert the last k-mer, too.
-        RELEASE_ALL_TAGS_SPIN_LOCK
-        if (found_tags) {
-            found_tags->insert(kmer);
+        if (this->get_count(kmer) < min_count) {
+            min_count = count;
         }
     }
+    return min_count;
 }
 
-//
-// consume_fasta_and_tag: consume a FASTA file of reads, tagging reads every
-//     so often.
-//
-
-// TODO? Inline in header.
-void
-Hashgraph::
-consume_fasta_and_tag(
-    std:: string const  &filename,
-    unsigned int	      &total_reads, unsigned long long	&n_consumed
-)
-{
-    IParser *	  parser =
-        IParser::get_parser( filename );
-
-    consume_fasta_and_tag(
-        parser,
-        total_reads, n_consumed
-    );
-
-    delete parser;
-}
-
-void
-Hashgraph::
-consume_fasta_and_tag(
-    read_parsers:: IParser *  parser,
-    unsigned int		    &total_reads,   unsigned long long	&n_consumed
-)
+BoundedCounterType Hashtable::get_max_count(const std::string &s)
 {
-    Read			  read;
-
-    // TODO? Delete the following assignments.
-    total_reads = 0;
-    n_consumed = 0;
+    KmerIterator kmers(s.c_str(), _ksize);
 
-    // Iterate through the reads and consume their k-mers.
-    while (!parser->is_complete( )) {
+    BoundedCounterType max_count = 0;
```

Comment:
https://en.wikipedia.org/wiki/Return_value_optimization it probably isn't that exciting in this case but for other/bigger/expensive to copy types it might be more interesting